File: python/querygraph/navigator.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime

from querygraph.cdif import CdifResource
from querygraph.croissant import CroissantDataset, Field, FileObject, RecordSet
from querygraph.did import DidDocument
from querygraph.odrl import Action, Policy, Rule


@dataclass(frozen=True)
class NavigatorInput:
    dataset_name: str
    description: str
    landing_page: str
    data_url: str
    creator: str
    agent_name: str


@dataclass(frozen=True)
class NavigatorOutput:
    generated_at: datetime
    croissant: dict
    cdif: dict
    did: DidDocument
    odrl: dict
    bundle: dict


class AiNavigator:
    def build(self, input: NavigatorInput) -> NavigatorOutput:
        did = DidDocument.new_oyd(
            f"{input.agent_name}:{input.creator}:{input.dataset_name}",
            input.agent_name,
        ).with_service_endpoint(input.landing_page)

        dataset_id = f"{input.landing_page.rstrip('/')}/#dataset"
        dataset = CroissantDataset(
            id=dataset_id,
            name=input.dataset_name,
            description=input.description,
            license="https://creativecommons.org/licenses/by/4.0/",
            creators=[input.creator],
            files=[
                FileObject(
                    id=f"{dataset_id}/file/source",
                    name="source-data",
                    content_url=input.data_url,
                    encoding_format="application/octet-stream",
                )
            ],
            record_sets=[
                RecordSet(
                    id=f"{dataset_id}/recordset/default",
                    name="default observations",
                    fields=[
                        Field(
                            "subject",
                            "sc:Text",
                            "Primary entity or observation subject",
                        ).semantic_type("https://schema.org/about"),
                        Field(
                            "value",
                            "sc:Text",
                            "Observed value, label, or narrative",
                        ).semantic_type("https://schema.org/value"),
                        Field(
                            "source",
                            "sc:URL",
                            "Evidence or provenance URL",
                        ).semantic_type("https://schema.org/citation"),
                    ],
                )
            ],
            keywords=["AI Navigator", "Croissant", "CDIF", "DID", "ODRL"],
        )

        policy = Policy(
            id=f"{dataset_id}/policy/default",
            target=dataset_id,
            assigner=did.id,
            permissions=[
                Rule(
                    action=Action.READ,
                    assignee="public",
                    constraint="attribution required",
                ),
                Rule(
                    action=Action.INDEX,
                    assignee=did.id,
                    constraint="local semantic indexing for AI Navigator",
                ),
            ],
            prohibitions=[
                Rule(
                    action=Action.DERIVE,
                    assignee="public",
                    constraint="no model training without separate agreement",
                )
            ],
        )
        odrl_json = policy.to_json_ld()
        cdif = CdifResource.from_croissant(
            dataset, input.landing_page, input.data_url
        ).with_odrl_policy(policy.id, odrl_json)

        croissant_json = dataset.to_json_ld()
        cdif_json = cdif.to_json_ld()
        generated_at = datetime.now(UTC)
        did_json = did.to_json()
        bundle = {
            "@context": {
                "schema": "https://schema.org/",
                "cr": "http://mlcommons.org/croissant/",
                "cdif": "https://cdif.codata.org/",
                "dcat": "http://www.w3.org/ns/dcat#",
                "dct": "http://purl.org/dc/terms/",
                "odrl": "http://www.w3.org/ns/odrl/2/",
                "querygraph": "https://querygraph.ai/ns#",
            },
            "@type": "querygraph:AiNavigatorSemanticBundle",
            "generatedAt": generated_at.isoformat().replace("+00:00", "Z"),
            "identity": did_json,
            "layers": {
                "semanticCroissant": croissant_json,
                "cdif": cdif_json,
                "did": did_json,
                "odrl": odrl_json,
            },
        }

        return NavigatorOutput(
            generated_at=generated_at,
            croissant=croissant_json,
            cdif=cdif_json,
            did=did,
            odrl=odrl_json,
            bundle=bundle,
        )

File: src/navigator.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use crate::{
    cdif::CdifResource,
    croissant::{CroissantDataset, Field, FileObject, RecordSet},
    did::DidDocument,
    odrl::{Action, Policy, Rule},
};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct NavigatorInput {
    pub dataset_name: String,
    pub description: String,
    pub landing_page: String,
    pub data_url: String,
    pub creator: String,
    pub agent_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct NavigatorOutput {
    pub generated_at: DateTime<Utc>,
    pub croissant: Value,
    pub cdif: Value,
    pub did: DidDocument,
    pub odrl: Value,
    pub bundle: Value,
}

#[derive(Debug, Default)]
pub struct AiNavigator;

impl AiNavigator {
    pub fn build(&self, input: NavigatorInput) -> NavigatorOutput {
        let dataset_id = format!("{}/#dataset", input.landing_page.trim_end_matches('/'));
        let dataset = CroissantDataset {
            id: dataset_id.clone(),
            name: input.dataset_name.clone(),
            description: input.description.clone(),
            license: "https://creativecommons.org/licenses/by/4.0/".to_string(),
            creators: vec![input.creator.clone()],
            files: vec![FileObject {
                id: format!("{dataset_id}/file/source"),
                name: "source-data".to_string(),
                content_url: input.data_url.clone(),
                encoding_format: "application/octet-stream".to_string(),
            }],
            record_sets: vec![RecordSet {
                id: format!("{dataset_id}/recordset/default"),
                name: "default observations".to_string(),
                fields: vec![
                    Field::new(
                        "subject",
                        "sc:Text",
                        "Primary entity or observation subject",
                    )
                    .semantic_type("https://schema.org/about"),
                    Field::new("value", "sc:Text", "Observed value, label, or narrative")
                        .semantic_type("https://schema.org/value"),
                    Field::new("source", "sc:URL", "Evidence or provenance URL")
                        .semantic_type("https://schema.org/citation"),
                ],
            }],
            keywords: vec![
                "AI Navigator".to_string(),
                "Croissant".to_string(),
                "CDIF".to_string(),
                "DID".to_string(),
                "ODRL".to_string(),
            ],
        };
        self.build_from_croissant(
            dataset,
            input.landing_page,
            input.data_url,
            input.creator,
            input.agent_name,
        )
    }

    pub fn build_from_croissant(
        &self,
        dataset: CroissantDataset,
        landing_page: impl Into<String>,
        access_service: impl Into<String>,
        creator: impl Into<String>,
        agent_name: impl Into<String>,
    ) -> NavigatorOutput {
        let landing_page = landing_page.into();
        let access_service = access_service.into();
        let creator = creator.into();
        let agent_name = agent_name.into();
        let did = DidDocument::new_oyd(
            format!("{}:{}:{}", agent_name, creator, dataset.name),
            agent_name,
        )
        .with_service_endpoint(landing_page.clone());

        let dataset_id = dataset.id.clone();
        let policy = Policy {
            id: format!("{dataset_id}/policy/default"),
            target: dataset_id.clone(),
            assigner: did.id.clone(),
            permissions: vec![
                Rule {
                    action: Action::Read,
                    assignee: "public".to_string(),
                    constraint: Some("attribution required".to_string()),
                },
                Rule {
                    action: Action::Index,
                    assignee: did.id.clone(),
                    constraint: Some("local semantic indexing for AI Navigator".to_string()),
                },
            ],
            prohibitions: vec![Rule {
                action: Action::Derive,
                assignee: "public".to_string(),
                constraint: Some("no model training without separate agreement".to_string()),
            }],
        };

        let croissant_json = dataset.to_json_ld();
        let odrl_json = policy.to_json_ld();
        let cdif_json = CdifResource::from_croissant(&dataset, landing_page, access_service)
            .with_odrl_policy(policy.id, odrl_json.clone())
            .to_json_ld();
        let generated_at = Utc::now();
        let bundle = json!({
            "@context": {
                "schema": "https://schema.org/",
                "cr": "http://mlcommons.org/croissant/",
                "cdif": "https://cdif.codata.org/",
                "dcat": "http://www.w3.org/ns/dcat#",
                "dct": "http://purl.org/dc/terms/",
                "odrl": "http://www.w3.org/ns/odrl/2/",
                "querygraph": "https://querygraph.ai/ns#"
            },
            "@type": "querygraph:AiNavigatorSemanticBundle",
            "generatedAt": generated_at,
            "identity": did,
            "layers": {
                "semanticCroissant": croissant_json,
                "cdif": cdif_json,
                "did": did,
                "odrl": odrl_json
            }
        });

        NavigatorOutput {
            generated_at,
            croissant: croissant_json,
            cdif: cdif_json,
            did,
            odrl: odrl_json,
            bundle,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builds_all_four_semantic_layers() {
        let output = AiNavigator.build(NavigatorInput {
            dataset_name: "Hazard vocabulary".to_string(),
            description: "Controlled vocabulary with multilingual technical terms".to_string(),
            landing_page: "https://querygraph.ai/datasets/hazards".to_string(),
            data_url: "https://querygraph.ai/datasets/hazards.csv".to_string(),
            creator: "QueryGraph".to_string(),
            agent_name: "AI Navigator".to_string(),
        });

        assert_eq!(output.croissant["@type"], "cr:Dataset");
        assert_eq!(output.cdif["@type"], "dcat:Dataset");
        assert!(output.did.id.starts_with("did:oyd:zQm"));
        assert_eq!(output.odrl["@type"], "odrl:Policy");
    }
}
